Scripts/Utilities/AssetDropUtils.cs

```diff
@@ -0,0 +1,132 @@
+        const string k_AttachScriptUndoLabel = "Add Script";
+        const string k_AssignMaterialUndoLabel = "Assign Material";
+        const string k_AssignPhysicMaterialUndoLabel = "Assign Physic Material";
+        const string k_AssignMaterialShaderUndoLabel = "Assign Material Shader";
+
+        internal static AnimationClip AttachAnimationClip(GameObject go, AssetData data)
+        {
+            var animation = go.GetComponent<Animation>();
+            if (animation == null)
+                animation = Undo.AddComponent<Animation>(go);
+
+            Undo.RecordObject(animation, k_AssignAnimationClipUndoLabel);
+            animation.clip = (AnimationClip)data.asset;
+
+            Undo.IncrementCurrentGroup();
+            return animation.clip;
+        }
+
+        internal static AudioClip AttachAudioClip(GameObject go, AssetData data)
+        {
+            var source = go.GetComponent<AudioSource>();
+            if (source == null)
+                source = Undo.AddComponent<AudioSource>(go);
+            
+            source.clip = (AudioClip)data.asset;
+
+            Undo.IncrementCurrentGroup();
+            return source.clip;
+        }
+
+        internal static VideoClip AttachVideoClip(GameObject go, AssetData data)
+        {
+            var player = go.GetComponent<VideoPlayer>();
+            if (player == null)
+                player = Undo.AddComponent<VideoPlayer>(go);
+
+            Undo.RecordObject(player, k_AssignVideoClipUndoLabel);
+            player.clip = (VideoClip)data.asset;
+
+            Undo.IncrementCurrentGroup();
+            return player.clip;
+        }
+
+        internal static Type AttachScript(GameObject go, AssetData data)
+        {
+            var script = (MonoScript)data.asset;
+            var type = script.GetClass();
+            Undo.AddComponent(go, type);
+            return type;
+        }
+
+        internal static Material AssignMaterial(GameObject go, AssetData data)
+        {
+            var renderer = go.GetComponent<Renderer>();
+
+            if (renderer != null)
+            {
+                Undo.RecordObject(go, k_AssignMaterialUndoLabel);
+                renderer.sharedMaterial = (Material)data.asset;
+                Undo.IncrementCurrentGroup();
+                return renderer.sharedMaterial;
+            }
+
+            return null;
+        }
+
+        internal static Shader AssignMaterialShader(GameObject go, AssetData data)
+        {
+            var renderer = go.GetComponent<Renderer>();
+
+            if (renderer != null)
+            {
+                Undo.RecordObject(go, k_AssignMaterialUndoLabel);
+                // this warns that we're leaking materials into the scene,
+                // and creates a new instance, but we don't want to change
+                // the shader on the shared material here.
+                var shader = (Shader)data.asset;
+                renderer.material.shader = shader;
```

Comment:
Do we want to create a duplicate material and change the shader there before assigning it?  Might as well put the new stuff in an expected place for the user

Reply:
We discussed this issue. The problem is that creating any kind of asset will cause frame drops. We decided a "good enough" solution is to leak the material into the scene and bypass the error.

Reply:
I wonder if we could 'pool' dummy materials in advance and use those - would still have to find a good time to make those however

Reply:
We'd have to actually write them to disk, and then clean up the unused ones afterward. And you'd have to either make an unreasonable amount, or you might run out.  I think the best course is to just wait for async asset database and make sure we can touch it without locking the UI

Reply:
Another option is to do the opposite: keep track of any "leaked" materials and write them all out when you close EXR. Since we needed a similar thing for the Poly workspace, it might make sense to set up an "egress" event where systems can do these one-time-on-shutdown actions